### DISCUSSION TITLE
Does View.center() need to be called several times ?

### DISCUSSION THREAD

**Discussion by Yarkane** [jimy-byerley/pymadcad#44] [Q&A] [answered]
Hi again,

This time, I'm trying to move the camera to a point with given coordinates. This point is a glm.fvec3 object with x, y and z values.

I'm calling, as seen in the documentation, the function View.center(vec) : 
https://pymadcad.readthedocs.io/en/latest/reference/rendering.html?highlight=center#madcad.rendering.View.center

But what I see is that my scene does only move a little bit. To get the perfect position I want, I need to call this function multiple times, depending on the initial position (if the camera was far away, it can be something like 400 calls before gettingthe position I need).

I can reduce the number of calls by calling in my loop center() then look() at the center of my scene :
```
            # Moving Camera
            for i in range(60): 
                view_3D.center(center_vec) #fvec3 of the position I want
                view_3D.look(barycenter) #fvec3 of the center of my object

            view_3D.adjust() # Out of the loop
```
But I still need between 40 to 100 calls to get the angle I need. 

Is that the expected behavior ? I thought by calling the function with, for example, a vector (10, 20, 30), we would immediately get the center of the camera being 10, 20, 30.

Thanks a lot for you attention ! :)

**Comment by jimy-byerley** [comment]
Hello again @Yarkane 
This should work as you expected, immediately placing the camera rotation center to the given position. As you can see in [`rendering.py`](https://github.com/jimy-byerley/pymadcad/blob/master/madcad/rendering.py#L897) the function is mostly a center assignment.

Are you sure you are not overriding your desired position afterward ? for instance if you reset in some way the camera center so that only the orientation remains then your added calls to `look()` and `adjust()` can be helping just because they are setting the orientation so that your object is in sight, and remains approximetely in sight after reset.
Or eventually are you using an  `Orbit` as navigation ? (It seems it has an issue with `look`)

**Comment by Yarkane** [comment, marked as answer]
Thanks for you answer, here are more information :

- It seems that you're right about the fact that `center()` place the camera to the correct position, as when I remove `look()`, the position of the camera is at the right place : calling `center()` a second time does not change anything, and I can rotate manually the camera so I get the view I wanted to have with `look()`.
- I don't know for sure what navigation object I'm using, but if the default tool is Orbit, then that's what I use, and as the issue seems to be on `look()` specifically it's coherent.

FYI, these two codes give equivalent results (that are incorrect, so) :
```
            # Moving camera
            view_3D.center(center_vec) # Center to the correct position

            # Zoom
           view_3D.adjust()

            # Rotate camera
            view_3D.look(barycenter) # Rotate the camera only a little bit and does not adjust the distance correctly
```

```
            # Moving camera
            view_3D.center(center_vec) # Center to the correct position

            # Zoom
            view_3D.adjust()

            # Rotate camera
            for i in range(30):
               view_3D.look(barycenter) 
             # Rotate the camera only a little bit and does not adjust the distance correctly even after n iterations
```

So, the only "working" implementation was the one on the beginning of the discussion : center -> look -> center -> look etc etc...
`adjust()` can be before or after `look()`, it does not affect the issue.

**Reply by jimy-byerley** [reply]
The default navigation object is `Turntable` so there shouldn't be any problem.

What exactly is the purpose of having 2 different center points `center_vec` and `barycenter` ?
I'm telling that because `look` is setting the navigation center to its given point, replacing the former one

**Reply by Yarkane** [reply]
Thank you for your answer,

I need to set the camera at a specific point, defined with `center_vec`, and look at an object, which has a `barycenter` value to get its position.
In the application context, `center_vec` would be the coordinates where a given photo has been taken, and I use `barycenter` only to make it look at the object in the scene.

If I only use `View.center(center_vec)`, the camera will have the right position, but will not look at the object.
If I only use `View.look(barycenter)`, the camera will look at the object but not be at the right place.
That's why I think it is necessary to call the two functions, event though I understand that `look` does more than just "rotating the camera to the item".

Am I wrong ? Is there another way to achieve this goal ? 

Thanks again for your time !

**Reply by jimy-byerley** [reply]
I understand better now ^^
Well this is not precisely what a `Turntable` navigation is for. and the issue comes from this fact.

The navigation object used by a `View` is managing the camera placement in the scene, this is also this object that decides how the camera is moved when receiving inputs from the mouse or the keyboard. By default, the navigation is a `Turntable` instance which makes the camera rotate around a center, and that center can translate.
This is meant to be convenient for the user using the GUI, but not always suited for other purpose.

`Turntable` can be thought to as a third person camera in a game, which rotation center is the player. When you rotarte the camera, you rotate it around the player, changing the camera position and orientation at the same time

The methods `View.look()`, `View.center()`, `View.adjust()` are dealing with that camera rotation center, but *not* with the camera position itself.
- `look` sets the camera rotation to see the point, then set the rotation point to the given point
- `center` only sets the rotation center to the given point

In your case, I understand that you want to control more precisely the camera position, just like if it was flying alone ?
If so, you can do either:
- use a `Turntable` as you do, but setting the camera-to-rotation-center distance to 0
- create a new navigation class, so you can give the camera view matrices you want to the `View`

I think I could also easily create a free-fly navigation class that could suit your need, allowing the user to set the camera matrix manually, but also to `look`, `center`, pan, zoom, etc without a rotation center :thinking:

**Reply by Yarkane** [reply]
Thank you for this reply which helped me a lot to understand how the navigation object works !
I solved my issue simply like this :

```
view.navigation.distance = 0
view.center(center_vec)
view.look(barycenter)
```

It does exactly what I need from what I see now. If I have another issue I will continue this discussion, but I'm happy with this for now.

Thanks again !

**Reply by jimy-byerley** [reply]
glad it helped !  you are welcome
